core/src/main/java/io/grpc/internal/StatsTraceContext.java

```diff
@@ -0,0 +1,220 @@
+    return Metadata.Key.of("grpc-census-bin", new Metadata.BinaryMarshaller<CensusContext>() {
+        @Override
+        public byte[] toBytes(CensusContext context) {
+          return context.serialize().array();
+        }
+
+        @Override
+        public CensusContext parseBytes(byte[] serialized) {
+          return censusContextFactory.deserialize(ByteBuffer.wrap(serialized));
+        }
+      });
+  }
+
+  /**
+   * Record the outgoing number of payload bytes as on the wire.
+   */
+  void wireBytesSent(long bytes) {
+    checkState(!callEnded, "already eneded");
+    wireBytesSent += bytes;
+  }
+
+  /**
+   * Record the incoming number of payload bytes as on the wire.
+   */
+  void wireBytesReceived(long bytes) {
+    checkState(!callEnded, "already eneded");
+    wireBytesReceived += bytes;
+  }
+
+  /**
+   * Record the outgoing number of payload bytes in uncompressed form.
+   *
+   * <p>The time this method is called is unrelated to the actual time when those byte are sent.
+   */
+  void uncompressedBytesSent(long bytes) {
+    checkState(!callEnded, "already ended");
+    uncompressedBytesSent += bytes;
+  }
+
+  /**
+   * Record the incoming number of payload bytes in uncompressed form.
+   *
+   * <p>The time this method is called is unrelated to the actual time when those byte are received.
+   */
+  void uncompressedBytesReceived(long bytes) {
+    checkState(!callEnded, "already ended");
+    uncompressedBytesReceived += bytes;
+  }
+
+  /**
+   * Record a finished all and mark the current time as the end time.
+   */
+  void callEnded(Status status) {
+    checkState(!callEnded, "already ended");
+    callEnded = true;
+    stopwatch.stop();
+    MetricName latencyMetric;
+    MetricName wireBytesSentMetric;
+    MetricName wireBytesReceivedMetric;
+    MetricName uncompressedBytesSentMetric;
+    MetricName uncompressedBytesReceivedMetric;
+    if (side == Side.CLIENT) {
+      latencyMetric = RpcConstants.RPC_CLIENT_ROUNDTRIP_LATENCY;
+      wireBytesSentMetric = RpcConstants.RPC_CLIENT_REQUEST_BYTES;
+      wireBytesReceivedMetric = RpcConstants.RPC_CLIENT_RESPONSE_BYTES;
+      uncompressedBytesSentMetric = RpcConstants.RPC_CLIENT_UNCOMPRESSED_REQUEST_BYTES;
+      uncompressedBytesReceivedMetric = RpcConstants.RPC_CLIENT_UNCOMPRESSED_RESPONSE_BYTES;
+    } else {
+      latencyMetric = RpcConstants.RPC_SERVER_SERVER_LATENCY;
+      wireBytesSentMetric = RpcConstants.RPC_SERVER_RESPONSE_BYTES;
+      wireBytesReceivedMetric = RpcConstants.RPC_SERVER_REQUEST_BYTES;
+      uncompressedBytesSentMetric = RpcConstants.RPC_SERVER_UNCOMPRESSED_RESPONSE_BYTES;
+      uncompressedBytesReceivedMetric = RpcConstants.RPC_SERVER_UNCOMPRESSED_REQUEST_BYTES;
+    }
+    censusContext
+        .with(RpcConstants.RPC_STATUS, new TagValue(status.getCode().toString()))
+        .record(MetricMap.builder()
+            .put(latencyMetric, stopwatch.elapsed(TimeUnit.MILLISECONDS))
```

Comment:
@dinooliva @sebright do you have a spec about what time unit should the latencies be reported in?